packages/commonwealth/server/controllers/server_threads_methods/archive_or_unarchive_thread.ts

```diff
@@ -0,0 +1,65 @@
+import { UserInstance } from '../../models/user';
+import { ServerThreadsController } from '../server_threads_controller';
+import { ThreadAttributes } from '../../models/thread';
+import { findAllRoles } from '../../util/roles';
+import { Op } from 'sequelize';
+
+const Errors = {
+  ThreadNotFound: 'Thread not found',
+  NotAdmin: 'Not an admin',
+};
+
+export type ArchiveOrUnarchiveThreadOptions = {
+  user: UserInstance;
+  threadId: string;
+  shouldArchive: boolean;
+};
+
+export type ArchiveOrUnarchiveThreadResult = ThreadAttributes;
+
+export async function __archiveOrUnarchiveThread(
+  this: ServerThreadsController,
+  { user, threadId, shouldArchive }: ArchiveOrUnarchiveThreadOptions
+): Promise<ArchiveOrUnarchiveThreadResult> {
+  const thread = await this.models.Thread.findOne({
+    where: {
+      id: threadId,
+    },
+  });
+  if (!thread) {
+    throw new Error(Errors.ThreadNotFound);
+  }
+  const userOwnedAddressIds = (await user.getAddresses())
```

Comment:
Would love to move some of this repeated logic into a helper function...

Reply:
Working on this. Just trying to make it somewhat flexible enough to use in other routes (though I won't over-engineer it at this point).